### ISSUE TITLE
e2e flaky: Connection refused errors [Errno 111] on jenkins

### ISSUE THREAD

**Issue by dchen1107** [kubernetes/kubernetes#23545]
We ran into a lot of resource leakage issues lately, which causes the test failed at the end:

http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gce-release-1.2/822/console
http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gce-release-1.2/820/console
http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gce-slow/3599/console
...

**Comment by ixdy** [comment]
820:
```
ERROR: gcloud crashed (CannotConnectToMetadataServerException): <urlopen error [Errno 111] Connection refused>
```

822:
```
ERROR: gcloud crashed (error): [Errno 111] Connection refused
```

Don't see anything like that for 3599, but I bet it was a similar issue.

**Comment by fejta** [comment]
Connection refused is almost certainly a jenkins cpu is overloaded issue

**Comment by gmarek** [comment]
cc @bprashanth @jlowdermilk

**Comment by jlowdermilk** [comment]
ci build of gcloud has retries to prevent crashing on connection refused from metadataserver. That was rolled back due to errors that look unrelated. I've re-installed ci build so these crashes should at least be less frequent. See #23544 for more context.

**Comment by fejta** [comment]
https://pantheon.corp.google.com/storage/browser/kubernetes-jenkins/logs/kubernetes-e2e-gce-release-1.0/4424/

**Comment by wojtek-t** [comment]
ping - this is still happening quite frequently, e.g.:
http://kubekins.dls.corp.google.com/view/Scalability/job/kubernetes-kubemark-5-gce/926/console

**Comment by gmarek** [comment]
@jlowdermilk - do you know if rolling back to 'old' gcloud version would solve the problem?

**Comment by fejta** [comment]
http://kubekins.dls.corp.google.com/job/kubernetes-e2e-gke-slow/2467/console

**Comment by fejta** [comment]
@gmarek @jlowdermilk @ixdy the core problem is that we are overloading the jenkins master. We need to start running jobs on slaves.

**Comment by jlowdermilk** [comment]
I think it's a combination of jenkins master overloading and GCE changes that made metadataserver a bit more flaky under load. Next week's release should see some improvement but we still need to run jobs on slaves.

**Comment by gmarek** [comment]
It's nearly constantly failing for Kubemarks:
```
06:35:28 gcloud docker push gcr.io/k8s-jenkins-kubemark/kubemark
06:35:28 ERROR: gcloud crashed (error): [Errno 111] Connection refused
```

**Comment by fejta** [comment]
@jlowdermilk and I have been talking to the GCE team about this. In parallel I have been working on a hacky server to cache requests to the metadata server on our jenkins VMs, but it looks like this will not work because of https://github.com/GoogleCloudPlatform/gcloud-golang/blob/master/compute/metadata/metadata.go#L208

**Comment by fejta** [comment]
Looks like I can work around this by adding multiple entries for metadata.google.internal to /etc/hosts

**Comment by fejta** [comment]
VICTORY

* http://kubekins.dls.corp.google.com/view/Submit%20Queue/job/kubernetes-e2e-gce/14631/
* https://pantheon.corp.google.com/storage/browser/kubernetes-jenkins/logs/kubernetes-e2e-gce/14631/

**Comment by wojtek-t** [comment]
@fejta - awesome!!!
Thanks a lot for working on it!

[even if this just reduces flakiness and not elimates it - what is happening currently is just a disaster]

**Comment by fejta** [comment]
BOOM! We are back up and merging: http://submit-queue.k8s.io/#/e2e 

The cache reduces load on the metadata server from ~20qps to ~0 qps (1 query per ten mins or so), and it also retries intermittent flakes refreshing the access token via the metadata server (since we know we're on gce it is safe to retry errno 111, which metadata.go and gcloud cannot do for perf reasons for regular users). So I expect this will elminate the errno 111 errors (no impact on other flakes)

**Comment by fejta** [comment]
So this was fixed all day until we migrated stuff over to docker, which caused problems because dns works differently. I'm rolling that back https://github.com/kubernetes/kubernetes/pull/24176

**Comment by wojtek-t** [comment]
@fejta - it seems that rollback didn't help - it is still failing a lot

**Comment by fejta** [comment]
Where are you still seeing errno 111s? Can you send me logs?

**Comment by wojtek-t** [comment]
Or maybe I looked for runs that are too old, e.g:
http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gce/14684/console

**Comment by yujuhong** [comment]
Just saw this on kubernetes-e2e-gke

https://console.cloud.google.com/storage/kubernetes-jenkins/logs/kubernetes-e2e-gke/7616/
https://console.cloud.google.com/storage/kubernetes-jenkins/logs/kubernetes-e2e-gke/7615/

**Comment by yujuhong** [comment]
@fejta why did you close the issue? Is there a fix or new issue that I missed?

**Comment by timstclair** [comment]
Also saw this, on GCE e2e builder: https://console.cloud.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/26735/kubernetes-pull-build-test-e2e-gce/43011/
```
2016/06/02 13:24:38 e2e.go:212: Running: build-release
ERROR: gcloud crashed (error): [Errno 111] Connection refused

If you would like to report this issue, please run the following command:
  gcloud feedback
2016/06/02 13:24:38 e2e.go:218: Error running build-release: exit status 1
2016/06/02 13:24:38 e2e.go:214: Step 'build-release' finished in 314.405288ms
2016/06/02 13:24:38 e2e.go:95: Error building. Aborting.
exit status 1
```

**Comment by bprashanth** [comment]
Yeah it's starting again: https://console.cloud.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/26734/kubernetes-pull-build-test-e2e-gce/43010/

```
rm -rf Godeps/_workspace # Just until we are sure it is gone
+ git clean -fdx
+ go run ./hack/e2e.go -v --build
2016/06/02 13:11:00 e2e.go:212: Running: build-release
ERROR: gcloud crashed (error): [Errno 111] Connection refused
```

@kubernetes/goog-testing

**Comment by spxtr** [comment]
It should be fixed. There was about half an hour where the metadata cache wasn't running on the master, but it should be fixed now.

**Comment by spxtr** [comment]
Ok, fixed now. Sorry about that.

**Comment by dchen1107** [comment]
I reopen this one. #26799 I manually restart deploy-gce jenkins job, but still failed with the same issue.

**Comment by ixdy** [comment]
Same issue. The Jenkins master VM restarted itself on Monday, and the metadata cache wasn't properly restarting itself at that time.

**Comment by dchen1107** [comment]
hmm, I manually restart, but failed; looks like after that @spxtr retried twice, failed one more time, then succeeded the second try. The problem is that if there is any magic behind it? Or just simply retry?

**Comment by spxtr** [comment]
I had to start the metadata cache on the master. We didn't notice the problem before because only the soak jobs run on the master.

**Comment by ixdy** [comment]
Can we get some sort of health check setup on the metadata cache? It is crashing far too frequently and we need to know before it fails jobs.

**Comment by spxtr** [comment]
Yes, that's a good idea.

**Comment by quinton-hoole** [comment]
@ixdy @fejta @spxtr @dchen1107 Drive-by comment.  Should we perhaps disable the metadata cache until it's more stable?  How much would that slow things down?

**Comment by spxtr** [comment]
@quinton-hoole It's not that the metadata cache isn't stable, it's that we don't check to make sure that it's up before starting a job and when Jenkins VMs restart we don't automatically start it. Without it, every job fails deterministically.

**Comment by ixdy** [comment]
@quinton-hoole I tried turning off the metadata cache a few weeks ago (using the metadata server directly), but the metadata server itself is still not stable enough for our use.

**Comment by wojtek-t** [comment]
Also happened to me - reopening.

**Comment by fgrzadkowski** [comment]
This is blocking a lot of PRs and merge queue is completely blocked. What are the instructions for restarting this cache?

**Comment by fgrzadkowski** [comment]
I have manually started metadata server of `pull-kubernetes-master`. We'll see if this'll help. For future reference I used the following command:

```
k8s.io/test-infra/jenkins/metadata-cache$ ./metadata-cache-control.sh remote_update pull-jenkins-master
```

**Comment by wojtek-t** [comment]
Can we document it somewhere?

**Comment by ixdy** [comment]
ideally we should fix the metadata server cache so it reliably restarts when the VM restarts.

**Comment by gmarek** [comment]
Documenting it before we're there sounds like a way to go.

**Comment by goltermann** [comment]
What remains before we can close this P0?  Is it just documenting how to restart the cache?

**Comment by spxtr** [comment]
Documenting and ensuring that the cache is running at all times. I don't think it's a P0 right now.

**Comment by fejta** [comment]
I'll update our internal documentation for this.
